snmp/datadog_checks/snmp/models.py

```diff
@@ -89,6 +89,19 @@ def __repr__(self):
         # type: () -> str
         return 'OID({!r})'.format(str(self))
 
+    def __hash__(self):
+        # type: () -> int
+        # Returns the hash of the repr of the underlying ObjectType
+        return hash(repr(self.as_object_type()))
```

Comment:
I think using `repr` can be costly to compute:

```
>>> repr(ObjectType(self._object_identity))
ObjectType(ObjectIdentity((1, 3, 6, 1, 4, 1, 30932, 1, 1, 2, 58, 2, 0)), <Null value object, tagSet <TagSet object, tags 0:0:5>, subtypeSpec <ConstraintsIntersection object, consts <SingleValueConstraint object, consts b''>>, encoding iso-8859-1, payload []>)
```

What about using `hash(self._parts)` ? 

I think this might be enough to cover deduplication of majority of cases (with value being a string oid like `1.3.6.1.4.1.30932.1.1.3.1.3.0`).

Example:

```
    def __hash__(self):
        # type: () -> int
        return hash(self._parts)

    def __eq__(self, other):
        # type: (object) -> bool
        if type(self) is type(other):
            other_oid = cast(OID, other)
            if self._parts is None or other_oid._parts is None:
                return False
            return hash(self._parts) == hash(other_oid._parts)
        else:
            return False
```

Reply:
Yes. I considered this option too, but the field `_parts` may unfortunately be `None` (see line 38).
After investigations, only `self.as_object_type()` is 100% safe for call.
I recognize that `repr` can be costy, but it is called only once when we read a profile. So I think the economy of SNMP calls worth it.

Reply:
Agree that it's currently called only once. But since OID class is quite widely used, I'm a bit worried that `__eq__` and `__hash__` could be called later elsewhere.

If we go this route, it might worth adding a comment about the costly nature of `__eq__` and `__hash__`.